### DISCUSSION TITLE
How to correctly set a GPIO pin high before the I2C and SSD1306 drivers are initialized

### DISCUSSION THREAD

**Discussion by zpm1066** [zephyrproject-rtos/zephyr#69310] [Q&A]
**Is your feature request related to a problem? Please describe.**
The request is how to correctly enable power on `I2C STEMMA` connector on a range of `Adafruit ESP32-S3` boards before any I2C and `SSD1306` (`ST7789`) display driver initialization.

The `Adafruit ESP32-S3` boards have on-board `I2C STEMMA` connector to allow I2C devices to be 
connected to it. In addition some of the boards have an on-board `ST7789` 135x240 TFT display. I'm trying to connect a I2C SSD1306 display and also use the on-board ST7789 display.

Power to both the I2C Stemma connector and the TFT is controlled by raising a GPIO pin high at a regulator. I2C STEMMA SDA and SCK have 10K pull-up resistors.

Under Arduino, the  initialization takes care of enabling power to I2C connector before the user's setup().

**Describe the solution you'd like**
I would like to enable power to the I2C STEMMA connector so that I can utilize both the on-board ST7789 display, as well as an I2C SSD1306 display (or any other I2C sensor) connected to the I2C STEMMA connector.

**Describe alternatives you've considered**
I've tried to use `gpio-hog` and also `regulator binding` controlled via a GPIO.

Here is the `gpio-hog` devicetree overlays that I used to set a GPIO pin high during initialization,

```
&gpio0 {
	status = "okay";

	i2c-tft-power {
		gpio-hog;
		gpios = <21 GPIO_ACTIVE_HIGH>;
		output-high;
	};
};
```

and the `regulator binding` with GPIO control

```
reg_tft_power: reg-tft-power {
	compatible = "regulator-fixed";
	regulator-name = "reg-adafruit";
	enable-gpios = <&gpio0 21 GPIO_ACTIVE_HIGH>;
	regulator-boot-on;
};
```

In my setup, I have a SSD1306 OLED display connected to the I22 STEMMA conenctor.

I’m able to confirm that the power to the display on the I2C STEMMA connector comes on when 
I use gpio-hog or the regulator binding. However, the display still doesn’t show anything.

From the “west espressif monitor”, I’ve confirmed that I2C is in a READY state but the 
SSD1306 failed to initialize. I do see 3.26V at the display connected via the I2C STEMMA 
and the SDA/SCK signals look fine. 

I2C is sending commands/data to the display correctly and there are no crashes on the ESP32-S3 and shell works fine.

I suspect that there may be a sequence issue, for example,  `gpio-hog` or `regulator binding` enabling the I2C power 
after SSD1306 but before I2C initialization.

Is there a way to check the order of I2C, SSD1306 and `gpio-hog` ( or `regulator binding`) priority in Zephyr?

Is `gpio-hog` or `regulator binding` the right way to turn on I2C power on initialization?

**Additional context**
Alternatively, Should I be using `SYS_INIT() with PRE_KERNEL` and set the I2C power 
GPIO high in early_init() as in the example below?

	`SYS_INIT(early_init, PRE_KERNEL_1, 0)`

**Comment by GeorgeCGV** [comment]
> Is there a way to check the order of I2C, SSD1306 and gpio-hog ( or regulator binding) priority in Zephyr?

Since Zephyr 3.5.0 there is `west build -t initlevels` to print out the init order.

A `SYS_INIT` is not a bad approach to deal with permanent board init.

**Comment by zpm1066** [comment]
@GeorgeCGV  Thanks.

In using `gpio-hog`

```
&gpio0 {
	status = "okay";
	i2c-tft-power {
		gpio-hog;
		gpios = <21 GPIO_ACTIVE_HIGH>;
		output-high;
	};
};
```

and from the `initlevels` output below, it looks like a   `__init_gpio_hogs_init: gpio_hogs_init(NULL`)

should set GPIO HIGH correctly init before I2C and SSD1306 gets initialized. 

I have verified that I'mgetting 3.26V at the I2C STEMMA connector and SDA/SCK levels look fine. Data is being pushed on to the I2C but SSD1306 is uninitialized.

 I'll try using `SYS_INIT(early_init, PRE_KERNEL_1, 0)`
 
What is the best way to get the GPIO 21 pin directly when calling from early_init?  Thanks.

```
[00:00:00.228,000] <err> i2c_esp32: I2C transfer error: -14
[00:00:00.228,000] <err> ssd1306: Failed to initialize device!
*** Booting Zephyr OS build zephyr-v3.5.0-4872-gcf14d4f1fd6d ***
[00:00:00.228,000] <inf> main: main_thread
Display device not ready
[00:00:00.228,000] <dbg> cfb: cfb_framebuffer_init: number of fonts 2
[00:00:00.317,000] <inf> display: Index[0] font dimensions  5x8
[00:00:00.317,000] <inf> display: Index[1] font dimensions  8x8
[00:00:00.317,000] <inf> display: Selected font: index[3]
[00:00:00.317,000] <inf> display: x_res 128, y_res 64, ppt 8, rows 8, cols 128
[00:00:00.317,000] <inf> display: Zephyr v3.5.99

[00:00:00.317,000] <inf> display: adafruit_esp32s3_tft_feather
uart:~$ device list
devices:
- rtc@60021000 (READY)
- gpio@60004800 (READY)
- gpio@60004000 (READY)
- trng@6003507c (READY)
- uart@60000000 (READY)
- uart@60038000 (READY)
- i2c@60013000 (READY)
- ssd1306@3c (DISABLED)
uart:~$ 

```

```
Successfully created esp32s3 image.
[409/410] cd /Users/quark11/Documents/2024_zephyr_dev/ssd13...128x64_dht_esp32s3_ref/build/zephyr/zephyr.elf --initlevels
EARLY
PRE_KERNEL_1
  __init___device_dts_ord_32: clock_control_esp32_init(__device_dts_ord_32)
  __init_init_mem_slab_obj_core_list: init_mem_slab_obj_core_list(NULL)
  __init_statics_init_pre: statics_init(NULL)
  __init___device_dts_ord_7: gpio_esp32_init(__device_dts_ord_7)
  __init___device_dts_ord_68: gpio_esp32_init(__device_dts_ord_68)
  __init___device_dts_ord_53: entropy_esp32_init(__device_dts_ord_53)
  __init___device_dts_ord_54: uart_esp32_init(__device_dts_ord_54)
  __init___device_dts_ord_57: serial_esp32_usb_init(__device_dts_ord_57)
  __init_uart_console_init: uart_console_init(NULL)
PRE_KERNEL_2
  __init_sys_clock_driver_init: sys_clock_driver_init(NULL)
POST_KERNEL
  __init_enable_logger: enable_logger(NULL)
  __init_enable_shell_uart: enable_shell_uart(NULL)
  __init_malloc_prepare: malloc_prepare(NULL)
  __init_k_sys_work_q_init: k_sys_work_q_init(NULL)
  __init_gpio_hogs_init: gpio_hogs_init(NULL)
  __init___device_dts_ord_70: i2c_esp32_init(__device_dts_ord_70)
  __init___device_dts_ord_72: ssd1306_init(__device_dts_ord_72)
  __init___device_dts_ord_8: dht_init(__device_dts_ord_8)
APPLICATION
  __init_lvgl_init: lvgl_init(NULL)
SMP
```

**Reply by GeorgeCGV** [reply]
It looks like the order is fine: gpios -> hogs -> i2c -> ssd1306. You wrote that the required pin is set before ssd1306 as well. It looks like i2c or ssd1396 driver issue.

**Reply by zpm1066** [reply]
Agreed. The order looks fine.

However, in the Espressif Monitor output, we do see that `SSD1306: Failed to initialise device!`

The `device list` confirms that I2C is ready but SSD1306 is disabled. Also, I2C is sending commands correctly and program isn’t crashing. It’s in a loop refreshing the display.

Is there a way to have gpio-hog in PRE KERNEL, so the I2T Power GPIO is set high much earlier?

If not, then I may try using SYS_INIT.

If I use something like SYS_INIT(early_init, PRE_KERNEL_1, 0), What is the best way to get the GPIO 21 pin directly when calling from early_init? How should I define it in the devicetree overlay? I don’t think I can have it under led GPIO section.


[[00:00:00.228,000] <err> i2c_esp32: I2C transfer error: -14
[00:00:00.228,000] <err> ssd1306: Failed to initialize device!
*** Booting Zephyr OS build zephyr-v3.5.0-4872-gcf14d4f1fd6d ***
[00:00:00.228,000] <inf> main: main_thread
Display device not ready
[00:00:00.228,000] <dbg> cfb: cfb_framebuffer_init: number of fonts 2
[00:00:00.317,000] <inf> display: Index[0] font dimensions  5x8
[00:00:00.317,000] <inf> display: Index[1] font dimensions  8x8
[00:00:00.317,000] <inf> display: Selected font: index[3]
[00:00:00.317,000] <inf> display: x_res 128, y_res 64, ppt 8, rows 8, cols 128
[00:00:00.317,000] <inf> display: Zephyr v3.5.99

[00:00:00.317,000] <inf> display: adafruit_esp32s3_tft_feather
uart:~$ device list
devices:
- rtc@60021000 (READY)
- gpio@60004800 (READY)
- gpio@60004000 (READY)
- trng@6003507c (READY)
- uart@60000000 (READY)
- uart@60038000 (READY)
- i2c@60013000 (READY)
- ssd1306@3c (DISABLED)

**Reply by GeorgeCGV** [reply]
> Is there a way to have gpio-hog in PRE KERNEL, so the I2T Power GPIO is set high much earlier?

The sda1306 driver allows to set a delay in init using `ready-time-ms` DT property. It is set to 10 milliseconds by default.

https://github.com/zephyrproject-rtos/zephyr/blob/dc8dff0fce08e3914751ea079aa7810971b192ec/drivers/display/ssd1306.c#L441

> If I use something like SYS_INIT(early_init, PRE_KERNEL_1, 0), What is the best way to get the GPIO 21 pin directly when calling from early_init? How should I define it in the devicetree overlay? I don’t think I can have it under led GPIO section.

It is ok to have it as led for a quick test. The `early_init` needs to get the gpio dt spec using macros. The blinky sample can be used as [example](https://github.com/zephyrproject-rtos/zephyr/blob/dc8dff0fce08e3914751ea079aa7810971b192ec/samples/basic/blinky/src/main.c).

**Reply by zpm1066** [reply]
#@GeorgeCGV  The SSD1306 connected to the I2C STEMMA connector works!  Thank you for your assistance. It's much appreciated.

I had to set `ready_time_ms`  in the `ssd1306` overlay to the display to come up. 

220 ms is the lowest value worked with the `esp32s3_tft_feather` and `esp32s3_feather` boards. I've gone with 250 ms.

```
&i2c0 {
	pinctrl-names = "default";
	status = "okay";
	clock-frequency = <I2C_BITRATE_STANDARD>;
	ssd1306: ssd1306@3c {
		compatible = "solomon,ssd1306fb";
		reg = <0x3c>;
		width = <128>;
		height = <64>;
		segment-offset = <0>;
		page-offset = <0>;
		display-offset = <0>;
		multiplex-ratio = <63>;
		segment-remap;
		com-invdir;
		prechargep = <0x22>;
		ready-time-ms = < 250 >;
	};
};
 #```

Now, I'm trying to get the ST7789 on the esp32s3_tft_feather to work but seem to be running into a similar issue with ST7789 TFT as I did with SSD1306. I would appreciate if you can suggest what may be happening.

I'm using the following overlay with `gpio-hog` to enable power to the TFT by setting GPIO I2C_TFT_Power high, as before. Also, setting the backlight.

The ST7789 CS, DC, RST and SCK/MOSI are set as per the Adafruit ESP32S3-TFT Feather pinouts.

```
// I2C_TFT_Power = GPIO21, TFT_Backlight = GPIO45
&gpio0 {
	status = "okay";
	i2c-tft-power {
		gpio-hog;
		gpios = <21 GPIO_ACTIVE_HIGH>, <45 GPIO_ACTIVE_HIGH>;
		output-high;
	};
};

&spi2 {
	compatible = "espressif,esp32-spi";
	status = "okay";
	cs-gpios = <&gpio0 7 GPIO_ACTIVE_LOW>; 

	st7789v: st7789v@0 {
		compatible = "sitronix,st7789v";
		label = "ST7789V";
		spi-max-frequency = <20000000>;
		reg = <0>;
		cmd-data-gpios = <&gpio0 39 GPIO_ACTIVE_LOW>; 
		reset-gpios = <&gpio0 40 GPIO_ACTIVE_LOW>;     
		width = <135>;
		height = <240>;
		x-offset = <52>;
		y-offset = <40>;
		vcom = <0x19>;
		gctrl = <0x35>;
		vrhs = <0x12>;
		vdvs = <0x20>;
		mdac = <0x00>;
		gamma = <0x01>;
		colmod = <0x05>;
		lcm = <0x2c>;
		porch-param = [0c 0c 00 33 33];
		cmd2en-param = [5a 69 02 01];
		pwctrl1-param = [a4 a1];
		pvgam-param = [D0 04 0D 11 13 2B 3F 54 4C 18 0D 0B 1F 23];
		nvgam-param = [D0 04 0C 11 13 2C 3F 44 51 2F 1F 1F 20 23];
		ram-param = [00 F0];
		rgb-param = [CD 08 14];
	};
};
```

The init order, `gpio_hogs` -> `spi_esp32_init` -> `st7789v_init` looks fine.

```
PRE_KERNEL_1
  __init___device_dts_ord_30: clock_control_esp32_init(__device_dts_ord_30)
  __init_init_mem_slab_obj_core_list: init_mem_slab_obj_core_list(NULL)
  __init_statics_init_pre: statics_init(NULL)
  __init___device_dts_ord_7: gpio_esp32_init(__device_dts_ord_7)
  __init___device_dts_ord_65: gpio_esp32_init(__device_dts_ord_65)
  __init___device_dts_ord_50: entropy_esp32_init(__device_dts_ord_50)
  __init___device_dts_ord_51: uart_esp32_init(__device_dts_ord_51)
  __init___device_dts_ord_54: serial_esp32_usb_init(__device_dts_ord_54)
  __init_uart_console_init: uart_console_init(NULL)
PRE_KERNEL_2
  __init_sys_clock_driver_init: sys_clock_driver_init(NULL)
POST_KERNEL
  __init_enable_logger: enable_logger(NULL)
  __init_enable_shell_uart: enable_shell_uart(NULL)
  __init_malloc_prepare: malloc_prepare(NULL)
  __init_k_sys_work_q_init: k_sys_work_q_init(NULL)
  __init_gpio_hogs_init: gpio_hogs_init(NULL)
  __init___device_dts_ord_70: spi_esp32_init(__device_dts_ord_70)
  __init___device_dts_ord_71: st7789v_init(__device_dts_ord_71)
  __init___device_dts_ord_8: dht_init(__device_dts_ord_8)
APPLICATION
  __init_lvgl_init: lvgl_init(NULL)
SMP
```

Here is the `Espressif Monitor` output. 
 
```
*** Booting Zephyr OS build zephyr-v3.5.0-4872-gcf14d4f1fd6d ***
[00:00:00.418,000] <inf> app: main
[00:00:00.419,000] <inf> app: Zephyr RTOS: 3.5.99

[00:00:00.419,000] <inf> app: adafruit_esp32s3_tft_feather
[00:00:00.419,000] <inf> app: Initialized ST7789V

uart:~$ device list
devices:
- rtc@60021000 (READY)
- gpio@60004800 (READY)
- gpio@60004000 (READY)
- trng@6003507c (READY)
- uart@60000000 (READY)
- uart@60038000 (READY)
- spi@60024000 (READY)
- ST7789V (READY)
```

Both SPI and ST7789 are in a READY state. The firmware is running fine with no errors.

If I switch to using the ST7789 overlay (after modifying the overlay to use different CS, DC, and RST pins) with an external standalone ST7789 TFT connected to the board pins, the ST7789 works without any issues.

I'm not sure what's going on. I feel I'm pretty close to getting the whole ST7789 to work.  

Any thoughts on what may be the issue? Thanks.

**Reply by GeorgeCGV** [reply]
Glad that it worked.

As for ST7789 make sure the max frequency is set correctly. Shouldn't it be 8 MHz for that display (`DT_FREQ_M(8)`)? Would be great to validate the existence of SPI data exchange.

Depending on the use case it might be necessary to provide it to `zephyr,display`.

Options like `width`, `height`, and others must be set accordingly.

**Reply by zpm1066** [reply]
#@GeorgeCGV 

I tried using 8Mz and also 1 MHz, it didn't make any difference.

 I think 20Mhz should be fine, as I have an identical ST7789 135x240 that runs fine when connected  to the TFT Feather board pin.
 The only change to the overlay was changing pins for CS, DC, and RST. The MOSI and SCK were the same as what the on-board ST7789 uses.
 
 I reviewed `spi_esp32_spim.c` but didn't see anything that stood out to be changed.
 

overlay snippet - as before zephyr display is set to st7789v:

```
 chosen {
		zephyr,display = &st7789v;
	};
```

and spi2 has the changed cs-gpios, cmd-data-gpios, and reset-gpios for the ST7789 using the Feather board pins.

```
&spi2 {
	compatible = "espressif,esp32-spi";
	status = "okay";
	cs-gpios = <&gpio0 9 GPIO_ACTIVE_LOW>; 

	st7789v: st7789v@0 {
		compatible = "sitronix,st7789v";
		label = "ST7789V";
		spi-max-frequency = <20000000>;
		reg = <0>;
		cmd-data-gpios = <&gpio0 10 GPIO_ACTIVE_LOW>;  
		reset-gpios = <&gpio0 8 GPIO_ACTIVE_LOW>;     
		width = <135>;
		height = <240>;
		x-offset = <52>;
		y-offset = <40>;
		vcom = <0x19>;
		gctrl = <0x35>;
		vrhs = <0x12>;
		vdvs = <0x20>;
		mdac = <0x00>;
		gamma = <0x01>;
		colmod = <0x05>;
		lcm = <0x2c>;
		porch-param = [0c 0c 00 33 33];
		cmd2en-param = [5a 69 02 01];
		pwctrl1-param = [a4 a1];
		pvgam-param = [D0 04 0D 11 13 2B 3F 54 4C 18 0D 0B 1F 23];
		nvgam-param = [D0 04 0C 11 13 2C 3F 44 51 2F 1F 1F 20 23];
		ram-param = [00 F0];
		rgb-param = [CD 08 14];
	};
};

```
